File: calc.py
#!/usr/bin/env python3
"""Démo en ligne de commande du moteur fiscal INEE2.0.

Exemples :
    python3 calc.py fr-ir 35000 --parts 1
    python3 calc.py fr-ir 60000 --parts 3 --couple
    python3 calc.py be-ipp 40000 --communal 0.08
    python3 calc.py lu-irpp 100000 --classe 2
    python3 calc.py de-est 50000
    python3 calc.py de-gmbh 100000 --hebesatz 4.9
"""

import argparse
import json

from engine import france, belgique, luxembourg, allemagne


def main() -> None:
    p = argparse.ArgumentParser(description="Moteur fiscal INEE2.0 (démo)")
    sub = p.add_subparsers(dest="cmd", required=True)

    a = sub.add_parser("fr-ir", help="France : impôt sur le revenu")
    a.add_argument("revenu", type=float)
    a.add_argument("--parts", type=float, default=1.0)
    a.add_argument("--couple", action="store_true")

    a = sub.add_parser("be-ipp", help="Belgique : IPP")
    a.add_argument("revenu", type=float)
    a.add_argument("--communal", type=float, default=0.0)

    a = sub.add_parser("lu-irpp", help="Luxembourg : IRPP")
    a.add_argument("revenu", type=float)
    a.add_argument("--classe", default="1")

    a = sub.add_parser("de-est", help="Allemagne : Einkommensteuer")
    a.add_argument("revenu", type=float)
    a.add_argument("--couple", action="store_true")

    a = sub.add_parser("de-gmbh", help="Allemagne : imposition GmbH")
    a.add_argument("benefice", type=float)
    a.add_argument("--hebesatz", type=float, default=4.0)

    args = p.parse_args()

    if args.cmd == "fr-ir":
        res = france.impot_revenu(args.revenu, parts=args.parts, couple=args.couple)
    elif args.cmd == "be-ipp":
        res = belgique.ipp(args.revenu, taux_communal=args.communal)
    elif args.cmd == "lu-irpp":
        res = luxembourg.irpp(args.revenu, classe=args.classe)
    elif args.cmd == "de-est":
        impot = allemagne.splitting(args.revenu) if args.couple else allemagne.einkommensteuer(args.revenu)
        res = {"einkommensteuer": impot,
               "soli": allemagne.solidaritaetszuschlag(impot, couple=args.couple)}
    elif args.cmd == "de-gmbh":
        res = allemagne.gmbh(args.benefice, hebesatz=args.hebesatz)

    print(json.dumps(res, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: src/calc.ts
#!/usr/bin/env node
/**
 * Démo CLI du moteur fiscal INEE2.0 (TypeScript).
 *
 *   npm run build && node dist/calc.js fr-ir 60000 --parts 3 --couple
 *   node dist/calc.js be-ipp 40000 --communal 0.08
 *   node dist/calc.js lu-irpp 100000 --classe 2
 *   node dist/calc.js de-est 100000 --couple
 *   node dist/calc.js de-gmbh 100000 --hebesatz 4.9
 */

import * as france from "./france.js";
import * as belgique from "./belgique.js";
import * as luxembourg from "./luxembourg.js";
import * as allemagne from "./allemagne.js";

function flag(args: string[], name: string): string | undefined {
  const i = args.indexOf(`--${name}`);
  return i >= 0 ? args[i + 1] : undefined;
}
function has(args: string[], name: string): boolean {
  return args.includes(`--${name}`);
}

function main(): void {
  const [cmd, valueRaw, ...rest] = process.argv.slice(2);
  const value = Number(valueRaw);
  let res: unknown;

  switch (cmd) {
    case "fr-ir":
      res = france.impotRevenu(value, Number(flag(rest, "parts") ?? 1), has(rest, "couple"));
      break;
    case "be-ipp":
      res = belgique.ipp(value, Number(flag(rest, "communal") ?? 0));
      break;
    case "lu-irpp":
      res = luxembourg.irpp(value, (flag(rest, "classe") ?? "1") as "1" | "1a" | "2");
      break;
    case "de-est": {
      const couple = has(rest, "couple");
      const impot = couple ? allemagne.splitting(value) : allemagne.einkommensteuer(value);
      res = { einkommensteuer: impot, soli: allemagne.solidaritaetszuschlag(impot, couple) };
      break;
    }
    case "de-gmbh":
      res = allemagne.gmbh(value, Number(flag(rest, "hebesatz") ?? 4.0));
      break;
    default:
      console.error("Commandes: fr-ir | be-ipp | lu-irpp | de-est | de-gmbh");
      process.exit(1);
  }
  console.log(JSON.stringify(res, null, 2));
}

main();
